docs/madr/decisions/011-multizone-envoy-admin.md

```diff
@@ -0,0 +1,120 @@
+# Multizone Envoy Admin Operations
+
+* Status: accepted
+
+Technical Story: https://github.com/kumahq/kuma/issues/5080
+
+## Context and Problem Statement
+
+Some time ago we introduced functionality that lets user execute Config Dump / Stats / Clusters from the CP.
+This is also available through Global CP in multi zone deployment.
+The problem is that Zone CP from given zone connects to only one instance of Global CP and admin operations are piped through KDS.
+In HA deployment, when you execute the Admin Operation request you don't know which instance you will use.
+
+## Considered Options
+
+* Discovery
+  * Config
+* Communication
+  * API Server
+  * Inter-CP Server
+
+## Pros and Cons of the Options
+
+We need to pass the request to a proper instance, for this we need to consider how we want to communicate between CP instances and how they will discover each other.
+
+## Discovery
+
+### Config
+
+We could gather all instances and IPs in one Config object (ConfigMap on Kubernetes).
+```yaml
+type: Config
+name: cp-instances
+config: |
+  {
+    "instances": [
+      {
+        "address": "192.168.1.100",
+        "apiServerHttpsPort": 5682,
+        "id": "<instance_id>"  
+      }
+    ]
+  }
+```
+
+Every CP instance would start a component that updates itself in the instances list if they are not on the list.
+Additionally, a leader would check the connection to all instances and remove them from the list. This way we can clear dead instances without using heartbeat, avoiding time based events, clock skews etc.
```

Comment:
We can try another approach here:
1. Every CP instance acquires a lock with the name `<instance_id>_<IP>`
2. CP Leader tries to acquire all locks that have the name `<instance_id>_<IP>` and it updates ConfigMap with the instance list.
3. When CP Leader acquires the lock `<instance_id>_<IP>` this means `<instance_id>` is terminated and CP Leader can remove it from the ConfigMap.

This approach allows us to:
* avoid active health checking of CP instances
* avoid conflicts when writing to ConfigMap (because only the CP leader performs write operations)

Reply:
How would you implement such locks on K8S without Postgres?

Reply:
You can have a callback that will run when you're elected https://github.com/kumahq/kuma/blob/master/pkg/plugins/bootstrap/k8s/plugin.go#L354

Reply:
>Which means that we create a separate ConfigMap for every instance just so we can have many locks?

Yes. Is it a problem?

> This implementation would be different than Universal I think and at that point, it might be better to just dump all the pods by the leader.

I don't think it will be different, we already have an interface:
```go
// LeaderCallbacks defines callbacks for events from LeaderElector
// It is guaranteed that each methods will be executed from the same goroutine, so only one method can be run at once.
type LeaderCallbacks struct {
	OnStartedLeading func()
	OnStoppedLeading func()
}
```
but it's not implemented on k8s. If we implement this on k8s, the code should be pretty much the same. But in that case, we delegate health checking to the leader election mechanism. Otherwise, we have to implement active health checking ourselves and it can be not super trivial.

Reply:
> Yes. Is it a problem?

Yeah, maybe it's not. Although this https://github.com/kumahq/kuma/pull/5292#discussion_r1019651124 would replace all of the above